File: docs/test-cases/token-smuggling.go
// Test case: token-smuggling (LLM01:2025)
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

var blockedDomains = []string{"paypal.com", "apple.com", "google.com"}

type translateReq struct {
	Text string `json:"text"`
}

func translateHandler(w http.ResponseWriter, r *http.Request) {
	var body translateReq
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}
	// BUG: body.Text is formatted into the prompt without unicode/norm.NFKC.
	// Homoglyph attacks (e.g. Cyrillic 'а' U+0430) pass through untouched.
	prompt := fmt.Sprintf("Translate: %s", body.Text)
	resp := callClaude("claude-haiku-4-5-20251001", prompt, 512)
	w.Write([]byte(resp))
}

func summarizeHandler(w http.ResponseWriter, r *http.Request) {
	review := r.URL.Query().Get("review")
	// BUG: no filter for RTL override (\u202E) or invisible chars (\u200B, \u200C, \uFEFF)
	// before they are embedded in the LLM prompt.
	prompt := fmt.Sprintf("Summarize this review: %s", review)
	resp := callClaude("claude-haiku-4-5-20251001", prompt, 256)
	w.Write([]byte(resp))
}

func checkURLHandler(w http.ResponseWriter, r *http.Request) {
	url := r.URL.Query().Get("url")
	// BUG: blocklist check runs on raw bytes with no NFKC normalization.
	// "раypal.com" (Cyrillic р) trivially bypasses strings.Contains.
	for _, d := range blockedDomains {
		if strings.Contains(url, d) {
			w.Write([]byte(`{"safe":false}`))
			return
		}
	}
	w.Write([]byte(`{"safe":true}`))
}

func callClaude(model, prompt string, maxTokens int) string { return "" }

func main() {
	http.HandleFunc("/translate", translateHandler)
	http.HandleFunc("/summarize", summarizeHandler)
	http.HandleFunc("/check-url", checkURLHandler)
	http.ListenAndServe(":8080", nil)
}

File: docs/test-cases/token-smuggling.rs
// Test case: token-smuggling (LLM01:2025)
use axum::{extract::Query, routing::get, Json, Router};
use serde::Deserialize;
use serde_json::{json, Value};

const BLOCKED_DOMAINS: &[&str] = &["paypal.com", "apple.com", "google.com"];

#[derive(Deserialize)]
struct SummarizeQ {
    text: String,
}

#[derive(Deserialize)]
struct UrlQ {
    url: String,
}

async fn summarize(Query(q): Query<SummarizeQ>) -> Json<Value> {
    // BUG: q.text is interpolated into the prompt with no
    // unicode_normalization::UnicodeNormalization::nfkc() pass. Homoglyphs
    // and compatibility forms reach the model unmodified.
    let prompt = format!("Summarize: {}", q.text);
    let out = call_claude("claude-haiku-4-5-20251001", &prompt, 256).await;
    Json(json!({ "summary": out }))
}

async fn translate(Query(q): Query<SummarizeQ>) -> Json<Value> {
    // BUG: no stripping of zero-width chars (\u{200B}, \u{200C}, \u{2060}, \u{FEFF})
    // or bidi overrides (\u{202E}) before prompt construction.
    let prompt = format!("Translate to English: {}", q.text);
    let out = call_claude("claude-haiku-4-5-20251001", &prompt, 512).await;
    Json(json!({ "translation": out }))
}

async fn check_url(Query(q): Query<UrlQ>) -> Json<Value> {
    // BUG: blocklist comparison on raw input — "раypal.com" (Cyrillic р, U+0440)
    // slips past contains() because the URL is never NFKC-normalized.
    for d in BLOCKED_DOMAINS {
        if q.url.contains(d) {
            return Json(json!({ "safe": false }));
        }
    }
    Json(json!({ "safe": true }))
}

async fn call_claude(_model: &str, _prompt: &str, _max_tokens: u32) -> String {
    String::new()
}

#[tokio::main]
async fn main() {
    let app = Router::new()
        .route("/summarize", get(summarize))
        .route("/translate", get(translate))
        .route("/check-url", get(check_url));
    axum::Server::bind(&"0.0.0.0:8080".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}
